x/ccv/provider/keeper/msg_server.go

```diff
@@ -309,16 +303,96 @@ func (k msgServer) SetConsumerCommissionRate(goCtx context.Context, msg *types.M
 	return &types.MsgSetConsumerCommissionRateResponse{}, nil
 }
 
-func (k msgServer) ConsumerModification(goCtx context.Context, msg *types.MsgConsumerModification) (*types.MsgConsumerModificationResponse, error) {
-	if k.GetAuthority() != msg.Authority {
-		return nil, errorsmod.Wrapf(types.ErrUnauthorized, "expected %s, got %s", k.GetAuthority(), msg.Authority)
+// CreateConsumer creates a consumer chain
+func (k msgServer) CreateConsumer(goCtx context.Context, msg *types.MsgCreateConsumer) (*types.MsgCreateConsumerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	consumerId := k.FetchAndIncrementConsumerId(ctx)
+
+	k.SetConsumerOwnerAddress(ctx, consumerId, msg.Signer)
+	k.SetConsumerChainId(ctx, consumerId, msg.ChainId)
+
+	if err := k.SetConsumerMetadata(ctx, consumerId, msg.Metadata); err != nil {
+		return &types.MsgCreateConsumerResponse{}, err
 	}
 
+	// initialization parameters are optional and hence could be nil
+	if msg.InitializationParameters != nil {
+		if err := k.SetConsumerInitializationParameters(ctx, consumerId, *msg.InitializationParameters); err != nil {
+			return &types.MsgCreateConsumerResponse{}, err
+		}
+	}
+
+	// power-shaping parameters are optional and hence could be null
+	if msg.PowerShapingParameters != nil {
+		if err := k.SetConsumerPowerShapingParameters(ctx, consumerId, *msg.PowerShapingParameters); err != nil {
+			return &types.MsgCreateConsumerResponse{}, err
+		}
+	}
+
+	if k.CanLaunch(ctx, consumerId) {
+		k.SetConsumerPhase(ctx, consumerId, Initialized)
+		k.PrepareConsumerForLaunch(ctx, consumerId, msg.InitializationParameters.SpawnTime)
+	} else {
+		k.SetConsumerPhase(ctx, consumerId, Registered)
+	}
+
+	return &types.MsgCreateConsumerResponse{ConsumerId: consumerId}, nil
+}
+
+// UpdateConsumer updates the record of a consumer chain
+func (k msgServer) UpdateConsumer(goCtx context.Context, msg *types.MsgUpdateConsumer) (*types.MsgUpdateConsumerResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	err := k.Keeper.HandleConsumerModificationProposal(ctx, msg)
+	consumerId := msg.ConsumerId
+
+	ownerAddress, err := k.Keeper.GetConsumerOwnerAddress(ctx, consumerId)
 	if err != nil {
-		return nil, errorsmod.Wrapf(err, "failed handling ConsumerModification proposal")
+		return &types.MsgUpdateConsumerResponse{}, errorsmod.Wrapf(types.ErrNoOwnerAddress, "cannot retrieve owner address %s", ownerAddress)
+	}
+
+	if msg.Signer != ownerAddress {
+		return &types.MsgUpdateConsumerResponse{}, errorsmod.Wrapf(types.ErrUnauthorized, "expected owner address %s, got %s", ownerAddress, msg.Signer)
+	}
+
+	if msg.PowerShapingParameters != nil && msg.PowerShapingParameters.Top_N > 0 && ownerAddress != k.GetAuthority() {
+		return &types.MsgUpdateConsumerResponse{}, errorsmod.Wrapf(types.ErrInvalidTransformToTopN,
+			"an update to a Top N chain can only be done if chain is owner is the gov module")
+	}
+
+	k.Keeper.SetConsumerOwnerAddress(ctx, consumerId, msg.NewOwnerAddress)
```

Comment:
What if `msg.NewOwnerAddress` is not set, i.e., is an empty field? Does the user need to always submit MsgUpdateConsumer with NewOwnerAddress set to the existing address?

Reply:
What if the current owner is the gov module and this message changes ownership to another address without changing TopN to zero?
The provider module should provide the following guarantees:
- only the existing owner can pass ownership to another account 
- a consumer with TopN > 0 MUST have the owner address set to the gov module account (this means that changing ownership from gov module entails automatically changing TopN to zero)

Reply:
>What if msg.NewOwnerAddress is not set, i.e., is an empty field? Does the user need to always submit MsgUpdateConsumer with NewOwnerAddress set to the existing address?

A transaction with an empty `NewOwnerAddress` would fail `MsgUpdateConsumer`s `ValidateBasic`. But we can add the same empty check here.

>Does the user need to always submit MsgUpdateConsumer with NewOwnerAddress set to the existing address?

Yes.

> What if the current owner is the gov module and this message changes ownership to another address without changing TopN to zero?

Can happen. The governance can "kill" this. But, adding a check here to guarantee that it can never happen.

> only the existing owner can pass ownership to another account

This is done in:
```
if msg.Signer != ownerAddress {
    return &types.MsgUpdateConsumerResponse{}, errorsmod.Wrapf(types.ErrUnauthorized, "expected owner address %s, got %s", ownerAddress, msg.Signer)
}	
```

> a consumer with TopN > 0 MUST have the owner address set to the gov module account (this means that changing ownership from gov module entails automatically changing TopN to zero)

This was not there. Added a check (i.e., `Top_N > 0 && OwnerAddress != Gov module`) on this and fail the transaction if the check succeeds, instead of setting the `Top_N` to 0.

Reply:
>> Does the user need to always submit MsgUpdateConsumer with NewOwnerAddress set to the existing address?

> Yes.

This is not ideal.

Reply:
Fixed mentioned issues.